app/ui/legacy/src/main/java/com/fsck/k9/activity/setup/AuthViewModel.kt

```diff
@@ -0,0 +1,269 @@
+import net.openid.appauth.AuthorizationService
+import net.openid.appauth.AuthorizationServiceConfiguration
+import net.openid.appauth.ResponseTypeValues
+import timber.log.Timber
+
+private const val KEY_AUTHORIZATION = "app.k9mail_auth"
+
+class AuthViewModel(
+    application: Application,
+    private val accountManager: AccountManager,
+    private val oauthCredentials: OAuthCredentials
+) : AndroidViewModel(application) {
+    private var authService: AuthorizationService? = null
+    private val authState = AuthState()
+
+    private var account: Account? = null
+
+    private lateinit var resultObserver: AppAuthResultObserver
+
+    private val _uiState = MutableStateFlow<AuthFlowState>(AuthFlowState.Idle)
+    val uiState: StateFlow<AuthFlowState> = _uiState.asStateFlow()
+
+    @Synchronized
+    private fun getAuthService(): AuthorizationService {
+        return authService ?: AuthorizationService(getApplication<Application>()).also { authService = it }
+    }
+
+    fun init(activityResultRegistry: ActivityResultRegistry, lifecycle: Lifecycle) {
+        resultObserver = AppAuthResultObserver(activityResultRegistry)
+        lifecycle.addObserver(resultObserver)
+    }
+
+    fun authResultConsumed() {
+        _uiState.update { AuthFlowState.Idle }
+    }
+
+    fun isAuthorized(account: Account): Boolean {
+        val authState = getOrCreateAuthState(account)
+        return authState.isAuthorized
+    }
+
+    private fun getOrCreateAuthState(account: Account): AuthState {
+        return try {
+            account.oAuthState?.let { AuthState.jsonDeserialize(it) } ?: AuthState()
+        } catch (e: Exception) {
+            Timber.e(e, "Error deserializing AuthState")
+            AuthState()
+        }
+    }
+
+    fun login(account: Account) {
+        this.account = account
+
+        viewModelScope.launch {
+            val config = findOAuthConfiguration(account)
+            if (config == null) {
+                _uiState.update { AuthFlowState.NotSupported }
+                return@launch
+            }
+
+            try {
+                startLogin(account, config)
+            } catch (e: ActivityNotFoundException) {
+                _uiState.update { AuthFlowState.BrowserNotFound }
+            }
+        }
+    }
+
+    private suspend fun startLogin(account: Account, config: OAuthConfiguration) {
+        val authRequestIntent = withContext(Dispatchers.IO) {
+            createAuthorizationRequestIntent(account.email, config)
+        }
+
+        resultObserver.login(authRequestIntent)
+    }
+
+    private fun createAuthorizationRequestIntent(email: String, config: OAuthConfiguration): Intent {
+        val serviceConfig = AuthorizationServiceConfiguration(
+            config.authorizationEndpoint.toUri(),
+            config.tokenEndpoint.toUri()
+        )
+
+        val applicationId = getApplication<Application>().packageName
+        val redirectUri = Uri.parse("$applicationId:/oauth2redirect")
```

Comment:
Shouldn't this be `$applicationId://oauth2redirect` with two slashes?
It seems the Microsoft Azure console does not allow adding a redirection URI with just one slash.

Reply:
I am trying to make this work with Office365/Outlook.
Are you maybe already working on this?
If not, I will open a new issue/MR for it.
But the blocker point will likely be to setup a k9mail application on Microsoft Azure, to have a dedicated client-id and allow the k9mail redirection URL.